honeybee_radiance/primitive.py

```diff
@@ -0,0 +1,486 @@
+                primitive should be included in the string. Default: True.
+        """
+        output = []
+
+        if include_dependencies:
+            for dep in self.dependencies:
+                output.append(self._to_radiance(dep, minimal))
+
+        if include_modifier and self.modifier.name != 'void':
+            output.append(self._to_radiance(self.modifier, minimal))
+        output.append(self._to_radiance(self, minimal))
+
+        return '\n'.join(output)
+
+    def to_dict(self):
+        """Translate this object to a dictionary."""
+        return {
+            "modifier": self.modifier.to_dict(),
+            "type": self.type,
+            "name": self.name,
+            "values": self.values,
+            "dependencies": [dep.to_dict() for dep in self._dependencies]
+        }
+
+    def duplicate(self):
+        """Get a copy of this object."""
+        return self.__copy__()
+
+    @staticmethod
+    def filter_dict_input(input_dict):
+        """Filter a dictionary of a Primitive to get modifier and dependency modules."""
+        try:  # ensure the putil module is imported, which imports all primitive modules
+            putil
+        except NameError:
+            import honeybee_radiance.putil as putil
+
+        # try to get modifier
+        if input_dict['modifier'] == 'void':
+            modifier = 'void'
+        else:
+            modifier = putil.dict_to_primitive(input_dict['modifier'])
+
+        if 'dependencies' not in input_dict: 
+            dependencies = []
+        else:
+            dependencies = [
+                putil.dict_to_primitive(dep) for dep in input_dict['dependencies']
+            ]
+
+        return modifier, dependencies
+
+    def __key(self):
+        """A tuple based on the object properties, useful for hashing."""
+        return (hash(self.modifier), self.type, self.name) + \
+            tuple(hash(tuple(vals)) for vals in self.values) + \
+            tuple(hash(dep) for dep in self._dependencies)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and \
+            self.__key() == other.__key()
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __copy__(self):
+        mod, depend = self._dup_mod_and_depend()
+        values_copy = [copy(line) for line in self._values]
+        return self.__class__(self.name, mod, values_copy, self._is_opaque, depend)
+    
+    def _dup_mod_and_depend(self):
+        """Duplicate this object's modifer and its dependencies."""
+        mod = None if isinstance(self._modifier, Void) else self._modifier.duplicate()
+        dependencies = []
+        for dep in self._dependencies:
+            try:
+                dependencies.append(dep.duplicate())  # it's a prmitive like a modifer
+            except AttributeError:
+                dependencies.append(dep)  # it's an immutable object like a string
```

Comment:
Technically, this should never happen. All the dependencies should be some sort of modifiers. Did we have test that was breaking because of this?

Reply:
No, I don't think this broke any of the tests but I remember you showing me a case where one of the dependencies was a `.cal` file and I assumed that would be coming here as a string. I'll see now if this causes any failing tests and let me know if I just wasn't understanding what a dependency could be.

Reply:
I can confirm that no tests failed. If you just confirm that my understanding of .cal files wasn't correct, I'll remove this.

Reply:
When it comes to dependency I think I only keep the modifiers that I find in the document in dependencies and ignore the rest so there will not be a case for .cal files, etc. The reason for that is I can't access the cal file even if I want to at that time and we can hope that the user puts the file in the right place at the right path.

Reply:
Good to know. I'll take out the try/except and merge this then. By the end of tomorrow, this library should be a full extension!